L107-csharp-nullable-reference-types.md

```diff
@@ -0,0 +1,169 @@
+the code generated without nullable support. Projects using C# 8.0 can take advantage
+of the nullable support if they so wish.
+
+### Code generation
+
+Options will be added to `protoc` and `grpc_csharp_plugin` to enable generating of code
+that supports nullable reference types. By default (without the new option) the code
+generated will be the same as today.
+
+Proposed name of option: *enable_nrt*
+
+e.g. 
+
+```bash
+protoc --plugin=protoc-gen-grpc=grpc_csharp_plugin \
+    --csharp_out=. \
+    --csharp_opt=enable_nrt \
+    --grpc_opt=enable_nrt \
+    myproto.proto
+```
+
+The details of the changes to the generated code are described below.
+
+### MSBuild integration
+
+`Grpc.Tools` will be changed to detect if a project has nullable enabled and
+automatically add the options to `protoc` and `grpc_csharp_plugin` if needed.
+
+A mechanism will be added to override this default behaviour. This will allow
+users to explicitly choose whether the code generated has nullable support or not.
+For example, it may be that the user wants to generate code that is compatible
+with other projects using an older compiler. This override will be done by setting
+the MSBuild property `gRPC_NullableReferenceTypes` to either `enable` or `disable`.
+
+## Rationale
+
+The _Nullable reference types_ feature was introduced in C# 8.0 in 2019. Supporting
+this language feature will give gRPC users information about null references in their
+code. For example:
+
+User's protocol buffers file:
+```proto
+message User {
+  string name = 1;
+  Address address = 2;
+}
+
+message Address {
+  string street = 1;
+  string city = 2;
+}
+```
+
+User's C# code:
+```c#
+public override Task<Empty> AddUser(User user, ServerCallContext context)
+{
+    AddressEntity entity = new
+    {
+        Street = user.Address.Street, // Possible null reference exception
+        City = user.Address.City,
+    };
+    _addressRepository.Add(entity);
+
+    // rest of the method
+}
+```
+
+Right now this will happily compile without warnings. If the generatad code had 
+added the annotations for nullable refernece types then the user would get a warning
+and would know that `User.Address` could be null.
+
+## Implementation
+
+There are three projects that need to be changed to fully support nullable reference types:
+- Protocol buffers C# code generation (Protocol buffers team)
+  - https://github.com/protocolbuffers/protobuf/tree/main/src/google/protobuf/compiler/csharp
+- gRPC C# plugin (gRPC team)
+  - https://github.com/grpc/grpc/tree/master/src/compiler
+- Grpc.Tools (gRPC team)
+  - https://github.com/grpc/grpc/tree/master/src/csharp/Grpc.Tools 
+
+It will require collaboration between the Protocol Buffers team and the gRPC team to
+coordinate the release of this feature.
+
+### Protocol buffers code generation
+
+Add the `enable_nrt` option, and when the enabled:
+- add `#nullable` directive
+- add annotations to Message fields as they can be null
```

Comment:
I didn't know that we know so precisely what exactly what would change (sounded like the reality will be more complicated that this), but if we do, that's great. 
FWIW, I think we don't need to mention exactly what changes would be required here, as long as we describe the general idea behind the changes and approx. what they would be. So I think we could at least qualify the statements here to not be as prescriptive?

Reply:
I've now qualified this by saying that other changes may become apparent during development.